File: src/sequence/accumulate.test.ts
import { describe, test, expect } from "vitest";
import { accumulate } from "./accumulate";

describe("accumulate", () => {
  test("adds value with plus factor", () => {
    expect(accumulate("plus", 10, 5)).toBe(15);
  });

  test("subtracts value with minus factor", () => {
    expect(accumulate("minus", 10, 5)).toBe(5);
  });

  test("works with zero base value", () => {
    expect(accumulate("plus", 0, 100)).toBe(100);
    expect(accumulate("minus", 0, 100)).toBe(-100);
  });

  test("works with zero value", () => {
    expect(accumulate("plus", 50, 0)).toBe(50);
    expect(accumulate("minus", 50, 0)).toBe(50);
  });

  test("handles negative results", () => {
    expect(accumulate("minus", 3, 10)).toBe(-7);
  });
});

File: accumulate_test.go
package aggrep

import "testing"

func TestAccumulate(t *testing.T) {
	t.Run("adds value with plus factor", func(t *testing.T) {
		if got := Accumulate(FactorPlus, 10, 5); got != 15 {
			t.Fatalf("got %v want 15", got)
		}
	})

	t.Run("subtracts value with minus factor", func(t *testing.T) {
		if got := Accumulate(FactorMinus, 10, 5); got != 5 {
			t.Fatalf("got %v want 5", got)
		}
	})

	t.Run("works with zero base value", func(t *testing.T) {
		if got := Accumulate(FactorPlus, 0, 100); got != 100 {
			t.Fatalf("got %v want 100", got)
		}
		if got := Accumulate(FactorMinus, 0, 100); got != -100 {
			t.Fatalf("got %v want -100", got)
		}
	})

	t.Run("works with zero value", func(t *testing.T) {
		if got := Accumulate(FactorPlus, 50, 0); got != 50 {
			t.Fatalf("got %v want 50", got)
		}
		if got := Accumulate(FactorMinus, 50, 0); got != 50 {
			t.Fatalf("got %v want 50", got)
		}
	})

	t.Run("handles negative results", func(t *testing.T) {
		if got := Accumulate(FactorMinus, 3, 10); got != -7 {
			t.Fatalf("got %v want -7", got)
		}
	})
}
